### ISSUE TITLE
Add missing host functions related to ledger header information

### ISSUE THREAD

**Issue by MonsieurNicolas** [stellar/rs-soroban-env#312]
Gaps:
* needed to write any kind of conditional based on "time"
    * ledger number
    * ledger close time
* needed to deal with core protocol ("host version") upgrades
    * protocol version

**Comment by leighmcculloch** [comment]
For the SDK side of this, if it is added at the same time, I think we should add a `Ledger` type that has these fns impl, where that ledger value is retrievable via the Env. 

```rust
impl Env {
    pub fn current_ledger(&self) -> Ledger
```

```rust
impl Ledger {
    pub fn number(&self) -> u32;
    pub fn close_time(&self) -> ?;
```

This pattern is the same pattern we use for `ContractData`.

**Comment by graydon** [comment]
On the host side I think we could do this either with some reserved magic-value storage keys or with something even lighter-weight like a constant symbol=>val map that a given host is populated with when it's constructed, representing its current (fixed-during-its-lifetime) execution context.

I think we should largely avoid providing host functions (or even SDK functions) with stellar-specific terminology; it would be hard to walk back / make-generic if we specialize too much, and relatively easy to keep generic / non-specialized.

**Comment by leighmcculloch** [comment]
I think we should avoid magically named contract data elements. They won't be naturally documented. They won't have types. I think SDK functions are appropriate, people will use this information in contracts and we should make it accessible.

In Solidity you get access to a `block` which has a `timestamp`, a `blockhash`, and a `blocknumber`, and so I don't think it is bespoke to provide similar information.

We could avoid Stellar terminology and use the term `timestamp` instead of `close_time`.

We could go further and use the term `block` instead of `ledger`, however `ledger` is particularly Stellar specific anyway, so it probably has little value in changing.

**Comment by jonjove** [comment]
We should also be realistic about the fact that some of the things we are building have stellar-specific meaning. To me that is both expected and important for getting a good level of integration. We have the host functions that get account thresholds and signers, for example. The token contract contains the notion of accounts in it. There will be host functions to move assets back and forth across the classic/smart boundary (which will only be usable by the token contract).

We should have a feature that we use to enable/disable the stellar-specific functionality. We could even move it into its own module to make it easy to include/exclude it.

**Comment by graydon** [comment]
Posted a version of this as described. Nothing fancy but enough to build on.

**Comment by leighmcculloch** [comment]
> needed to deal with core protocol ("host version") upgrades
> * protocol version

I'm very interested in the fact that from best I can tell Solidity doesn't provide a way to determine the EVM version inside a contract. The following Solidity documentation section for the available global variables is what I'm basing this on: https://docs.soliditylang.org/en/latest/units-and-global-variables.html?highlight=block#special-variables-and-functions

If there's no way to determine that within an Ethereum contract, I wonder what we could do differently to avoid contracts coding around protocol versions, and therefore not have a need to include this.

**Comment by MonsieurNicolas** [comment]
We need access to the protocol version at least from host functions and native contracts that need to switch behavior when we rev protocol version.
I don't know if normal contracts can be written 100% without access to the protocol version. I imagine that most people would not need it (and just code and deploy against "current"). It may be useful for people that want to deploy their contracts ahead of a protocol change (similar to native contracts) or kill their contracts with certain protocol changes for example (I'm thinking rollups need perfect parity between L1 and L2 - so it is a hard requirement to have L1.version == L2.version).

**Comment by jonjove** [comment]
@leighmcculloch @MonsieurNicolas should this be closed given #330 was merged, or is there more work to do?

**Comment by leighmcculloch** [comment]
This can be closed because the env exposes the data now in host functions. The SDK still needs to use those host functions to expose the same data in the SDK. That work is tracked in https://github.com/stellar/rs-soroban-sdk/issues/453.
